### ISSUE TITLE
Tracking issue for libc 0.3

### ISSUE THREAD

**Issue by joshtriplett** [rust-lang/libc#3248]
This issue tracks changes we want to make in libc 0.3.

- [ ] musl 1.2

**Comment by bossmc** [comment]
On the LFS64 symbol removal point (#2935) the current PR is attempting to achieve rust crate API compatibility (by aliasing LFS64 things to their non-LFS64 equivalents), hiding the effective ABI break in musl libc (it's not a real break since those symbols never "really" existed, but they existed enough for the libc crate to wind up depending on them 😞).  If 0.3 is going to be a breaking change we could do a simpler change where we simply stop all the LFS64 functions and types from the libc crate (which is arguably closer to the musl libc API).

My expectation was that we'd merge the API-compatible change in a 0.2.x (and maybe then deprecate the aliases?) and remove them "later" in a 0.3 release, but if 0.3 is imminent, maybe we just drop the items completely?

Note that the rust libc crate has to do "something" here to work with musl 1.2.4+ (@wesleywiser moved rustc's self-contained version to 1.2.3 since that's ABI-compatible with 1.1.24 but 1.2.4 breaks the crate's use of the ABI).

**Comment by joshtriplett** [comment]
@bossmc Figuring out how to deal with LFS64 and time64 has been one of the two biggest blocking issues for this. (The other is having unnamed fields of struct and union types.)

We *could* just try to match the API of musl as closely as possible, but that just pushes the problem onto whoever *uses* libc.

Personally, I would like to minimize the degree to which people have to write a pile of `cfg`s in order to use the 64-bit functions. `libc::foo64` seems preferable to `#[cfg(...)] use libc::foo as foo64; #[cfg(not(...))] use libc::foo64;` (with corresponding complexity for every other function as well as off_t).

Is there some reasonable way we could *generally* make the same function names and types available on every platform that has 64-bit support, so that people only need `cfg` if they want to support platforms that *don't* have 64-bit support?

**Comment by lvella** [comment]
So, the changes for 64 bits version of time and file offset function is just for `musl`? Can't 32 bit `glibc` be included in the package?

**Comment by JohnTitor** [comment]
> Fork a libc-0.2 branch from master on which development for libc 0.2 will continue.

Created the [`libc-0.2` branch](https://github.com/rust-lang/libc/tree/libc-0.2). Once the musl 1.2 PR (or any other breaking changes) gets merged, I'll update the bors config and docs.

**Comment by ydirson** [comment]
> > Fork a libc-0.2 branch from master on which development for libc 0.2 will continue.
> 
> Created the [`libc-0.2` branch](https://github.com/rust-lang/libc/tree/libc-0.2). Once the musl 1.2 PR (or any other breaking changes) gets merged, I'll update the bors config and docs.

Great to see 0.3 starting to take shape!
Is there an ETA for the `breakage-candidate`s to land in main branch, and for a 0.3.0?  Not asking for a firm date, my question is really "should I invest time to get a feature-guarded version of #3367 / #3201 into 0.2, or is 0.3 just right off the corner?"

**Comment by Xeonacid** [comment]
Will PR submitted to `libc-0.2`  get merged into main for 0.3? Should I make two PRs basing both `libc-0.2` and `main` if I want it available for both 0.2 and 0.3?

**Comment by Amanieu** [comment]
You should make 2 separate PRs.

**Comment by safinaskar** [comment]
Obviously, next libc release should use semver trick ( https://github.com/dtolnay/semver-trick ), because libc types widely used in public interfaces

**Comment by Amanieu** [comment]
I've updated the tracking issue to talk about libc 1.0 instead of 0.3.

**Comment by safinaskar** [comment]
No pull requests was accepted in one whole month. So, our 1.0 work is stalled

**Comment by Xuanwo** [comment]
It seems the reviewer for this repository is overwhelmed: there are 53 pull requests awaiting review, and almost all of them are for @JohnTitor. So do you need help? I'm willing to help push forward.

**Comment by tgross35** [comment]
Our branches unfortunately seem to have been diverging - some PRs target `main`, some target `libc-0.2`, and some both. This isn't very sustainable - it means the same reviews are happening in multiple places, and we are building up to either an ugly merge or surprise regressions whenever 1.0 releases.

So, please always target `main` with new PRs. If the PR is nonbreaking, comment `@rustbot label +stable-nominated` to indicate we may want it in a release before 1.0. After merge, it will then be cherry picked to the 0.2 branch (I have been doing these in batches and intend to continue to do so), assuming it applies cleanly, in which case the label will be changed to `stable-applied`.

I have updated documentation to reflect this.

**Comment by tgross35** [comment]
One other notable thing I think we should definitely do is make all padding or reserved fields `MaybeUninit`, as discussed here https://github.com/rust-lang/libc/issues/1453. This ensures the data can't be read by accident either on the user side or on our side via e.g. `#[derive(PartialEq)]`.

**Comment by safinaskar** [comment]
@tgross35 and others: I will say again: we definitely should do semver trick ( https://github.com/dtolnay/semver-trick ) to minimize breakage

**Comment by tgross35** [comment]
> [@tgross35](https://github.com/tgross35) and others: I will say again: we definitely should do semver trick ( https://github.com/dtolnay/semver-trick ) to minimize breakage

We still don't have a timeline for 1.0 (we'll have a better idea here after the time_t changes get in) but I think we will be able to do this when the time comes.

**Comment by highjeans** [comment]
Hi,

I saw this issue was linked as part of the Google Summer of Code 2025 project Idea list, I am very interested in working on this. I had a few questions about this project:
On the idea list, it mentioned a release of libc 0.3, however on this issue, it mentions the release of libc 1.0. Will the GSoC project be released as 0.3 or 1.0 (this is just for me to write about it in my proposal)?
Other than reading through some code, the CONTRIBUTING.md file, and the proposal guide, what would you guys think is good for me to prepare for working on this for GSoC?

Thanks,
Vivan Waghela

**Comment by tgross35** [comment]
I see you re-asked on the Zulip, we should use that to discuss GSoC https://rust-lang.zulipchat.com/#narrow/channel/421156-gsoc/topic/Idea.3A.20modernize.20the.20libc.20crate. I replied there, I'll hide these comments to keep discussion focused.

**Comment by tgross35** [comment]
There are some polls about changes for 1.0 ongoing on Zulip: https://rust-lang.zulipchat.com/#narrow/channel/219381-t-libs/topic/Questions.20and.20polls.20about.20.60libc.60.201.2E0/with/511533169

**Comment by lygstate** [comment]
I think `wchar_t` should be placed as  `core::ffi:c_wchar_t`.
The `core::ffi::c_ssize_t` and `core::ffi::c_size_t` should be stablized and used.
